src/rebase.c

```diff
@@ -975,10 +978,45 @@ static int rebase_commit__create(
 		message = git_commit_message(current_commit);
 	}
 
-	if ((error = git_commit_create(&commit_id, rebase->repo, NULL, author,
-		committer, message_encoding, message, tree, 1,
-		(const git_commit **)&parent_commit)) < 0 ||
-		(error = git_commit_lookup(&commit, rebase->repo, &commit_id)) < 0)
+	/* this error will be cleared by the signing process, but should be set
+	 * to signal the unsigned commit create process if we are not going to sign */
+	error = GIT_PASSTHROUGH;
+	if (rebase->options.signing_cb) {
+		if ((error = git_commit_create_buffer(&commit_content, rebase->repo, author, committer,
+				message_encoding, message, tree, 1, (const git_commit **)&parent_commit)) < 0)
+			goto done;
+
+		git_error_clear();
+		if ((error = rebase->options.signing_cb(&commit_signature, &signature_field,
+				git_buf_cstr(&commit_content), rebase->options.payload)) < 0 &&
+				error != GIT_PASSTHROUGH) {
+			if (git_error_last() == NULL)
+				git_error_set(GIT_ERROR_CALLBACK, "commit signing_cb failed");
+			goto done;
+		}
+
+		if (error != GIT_PASSTHROUGH) {
+			if (git_buf_is_allocated(&signature_field)) {
+				assert(git_buf_contains_nul(&signature_field));
+				signature_field_as_string = git_buf_cstr(&signature_field);
+			}
+
+			assert(git_buf_is_allocated(&commit_signature));
+			assert(git_buf_contains_nul(&commit_signature));
+			if ((error = git_commit_create_with_signature(&commit_id, rebase->repo,
+					git_buf_cstr(&commit_content), git_buf_cstr(&commit_signature),
+					signature_field_as_string)))
+				goto done;
+		}
+	}
+
+	/* if we skipped signing, create the commit normally */
+	if (error == GIT_PASSTHROUGH &&
+		(error = git_commit_create(&commit_id, rebase->repo, NULL, author, committer,
+			message_encoding, message, tree, 1, (const git_commit **)&parent_commit)) < 0)
+		goto done;
```

Comment:
The flow would be simplified by quite a bit if you simply always called `git_commit_create_with_signature`, leaving `commit_content` and `commit_signature` fields as `NULL` in case there was no signing callback provided or if it returned `GIT_PASSTHROUGH`.

Reply:
This ended up being much simpler, though I needed to tweak git_commit_create_with_signature to support NULL signatures.